Lib/test/test_ctypes/test_dlerror.py

```diff
@@ -0,0 +1,88 @@
+import os.path
+import sys
+import unittest
+from ctypes import CDLL
+
+FOO_C = r"""
+#include <stdio.h>
+
+/* This is a 'GNU indirect function' (IFUNC) that will be called by
+   dlsym() to resolve the symbol "foo" to an address. Typically, such
+   a function would return the address of an actual function, but it
+   can also just return NULL.  For some background on IFUNCs, see
+   https://willnewton.name/uncategorized/using-gnu-indirect-functions/
+
+   Adapted from Michael Kerrisk's answer: https://stackoverflow.com/a/53590014
+*/
+
+asm (".type foo, @gnu_indirect_function");
+
+void *foo(void)
+{
+    fprintf(stderr, "foo IFUNC called\n");
+    return NULL;
+}
+"""
+
+
+@unittest.skipUnless(sys.platform.startswith('linux'),
+                     'Test only valid for Linux')
+class TestNullDlsym(unittest.TestCase):
+    """GH-126554: Ensure that we catch NULL dlsym return values
+
+    In rare cases, such as when using GNU IFUNCs, dlsym(),
+    the C function that ctypes' CDLL uses to get the address
+    of symbols, can return NULL.
+
+    The objective way of telling if an error during symbol
+    lookup happened is to call glibc's dlerror() and check
+    for a non-NULL return value.
+
+    However, there can be cases where dlsym() returns NULL
+    and dlerror() is also NULL, meaning that glibc did not
+    encounter any error.
+
+    In the case of ctypes, we subjectively treat that as
+    an error, and throw a relevant exception.
+
+    This test case ensures that we correctly enforce
+    this 'dlsym returned NULL -> throw Error' rule.
```

Comment:
Is this summarized on the issue? if so, we don't need to be extra verbose here. Tests are usually not that descriptive in what they do, except if there are some real needs.

Reply:
Eh, I'm personally OK with it here. This is a complicated issue and equally complicated test case, I'd rather more info on it than not enough.

Reply:
Ok, I can buy that argument. Though I'd like to know whether this could garble the output in verbose mode (this is something we don't want) in case of a failure/success. If this isn't the case, then I'm also ok (but the description is alsmot longer than the test I think?)

Reply:
I'll double-check. I do remember seeing docstrings in some of the C API tests, I think.